outlines/models/base.py

```diff
@@ -0,0 +1,84 @@
+from abc import ABC, abstractmethod
+from typing import Literal, Union
+
+
+class ModelTypeAdapter(ABC):
+    """Base class for all model type adapters.
+
+    A type adapter instance must be given as a value to the `type_adapter`
+    attribute when instantiating a model.
+    The type adapter is responsible for formatting the input and output types
+    passed to the model to match the specific format expected by the
+    associated model.
+
+    """
+
+    @abstractmethod
+    def format_input(self, model_input):
+        """Format the user input to the expected format of the model.
+
+        For API-based models, it typically means creating the `messages`
+        argument passed to the client. For local models, it can mean casting
+        the input from str to list for instance.
+        This method is also used to validate that the input type provided by
+        the user is supported by the model.
+
+        """
+        ...
+
+    @abstractmethod
+    def format_output_type(self, output_type):
+        """Format the output type to the expected format of the model.
+
+        For API-based models, this typically means creating a `response_format`
+        argument. For local models, it means formatting the logits processor to
+        create the object type expected by the model.
+
+        """
+        ...
+
+
+class Model(ABC):
+    """Base class for all models.
+
+    This class defines a shared `__call__` method that can be used to call the
+    model directly.
+    All models inheriting from this class must define a `type_adapter`
+    attribute of type `ModelTypeAdapter`. The methods of the `type_adapter`
+    attribute are used in the `generate` method to format the input and output
+    types received by the model.
+
+    """
+
+    model_type: Union[Literal["api"], Literal["local"]]
```

Comment:
TIL about `typing.Litteral` in Python, nice! Are they like _"atoms"_ in Scheme, and a way to build kind of enums?

Reply:
Yes! But they are going away as we're going to keep the `LocalModel` and `ApiModel` types

Reply:
`Union[Literal["api"], Literal["local"]]` is equivalent to `Literal["api", "local"]` fyi